### PR TITLE
8273104: Refactoring option parser for UL

### PR DESCRIPTION
**Pull request by YaSuenag**
As we discussed in PR #4885 and in [its CSR](https://bugs.openjdk.java.net/browse/JDK-8271188), we want to introduce `foldmultilines` to stdout/err UL output. However we have no chance to configure `LogOutput` for stdout/err (`LogStdoutOutput`/`LogStderrOutput`) becasuse they will be instantiated statically in logFileStreamOutput.cpp.

So we need to refactor UL option parser to propagate options to all extended classes of `LogOutput`.

* Move `parse_option()` to `LogOutput` from `LogFileOutput`, then it becames public member.
* Introduce `set_option()` in `LogOutput` as a pure virtual member to apply options to each log output, then it is implemented in both `LogFileStreamOutput` and `LogFileOutput`.
* Move both `FoldMultilinesOptionKey` and `_fold_mulilines` in `LogFileStreamOutput` to private member because they are no longer to need to be public.

I want to send PR to introduce `foldmultilines` to stdout/err like https://github.com/YaSuenag/jdk/compare/ul-refactoring...YaSuenag:foldmultilines-for-stdout after this.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273104](https://bugs.openjdk.java.net/browse/JDK-8273104): Refactoring option parser for UL

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to e1f098e04bf384e438c216d8fc216f8cf1895788
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5293/head:pull/5293` \
`$ git checkout pull/5293`

Update a local copy of the PR: \
`$ git checkout pull/5293` \
`$ git pull https://git.openjdk.java.net/jdk pull/5293/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5293`

View PR using the GUI difftool: \
`$ git pr show -t 5293`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5293.diff">https://git.openjdk.java.net/jdk/pull/5293.diff</a>

</details>

### CODE REVIEW

**Review by iklam — APPROVED** [review]
This refactoring looks good to me. Thanks for doing it.

**Comment by YaSuenag** [comment]
PING: Can I get second reviewer?

File: src/hotspot/share/logging/logOutput.cpp [hunk 1]
```diff
@@ -339,3 +339,41 @@ void LogOutput::update_config_string(const size_t on_level[LogLevel::Count]) {
...
+    }
+
+    char* key = pos;
+    char* value_str = equals_pos + 1;
+    *equals_pos = '\0';
+    julong errstream_count_before = errstream->count();
+    success = set_option(key, value_str, errstream);
+    if (!success) {
+      if (errstream->count() > errstream_count_before) {
+        errstream->print_cr("Invalid option '%s' for log output (%s).", key, name());
```
Comment (dholmes-ora) [inline src/hotspot/share/logging/logOutput.cpp:370]: I don't quite follow this. If the error count has increased then this was a known option with a bad value that has already been reported in set_option. Here we should be catching an unknown option that wasn't recognized by set_option, in which case the error count should be unchanged no?

**Review by dholmes-ora — COMMENTED** [review]
Hi Yasumasa,

I have one query below about a possible correctness issue with the error messages.

IIUC these dynamic changes to the options will only apply after argument processing is complete - so anything logged prior to that will not be affected. That isn't an issue for options that don't directly change the output but for FoldMultiLines it means we would not do the folding initially. That probably isn't a practical concern given what logging can happen prior to argument processing ... or am I mistaken in thinking that UL can be used prior to that point?

Thanks,
David

**Comment by YaSuenag** [comment]
Thanks David for the comment!

> IIUC these dynamic changes to the options will only apply after argument processing is complete

All of output options can't be changed dynamically. We can see following error when we attempt to do so.
As you said UL for stdout/err can be used before argument processing, however file output isn't so. All of argument processing can be happened at `LogFileOutput::initialize()` and it would call `parse_options()`.

```
$ jcmd 62838 VM.log list
  <snip>
Log output configuration:
 #0: stdout all=warning uptime,level,tags
 #1: stderr all=off uptime,level,tags
 #2: file=test.log all=info uptime,level,tags filecount=5,filesize=20480K,async=false (reconfigured)

$ jcmd 62838 VM.log output=#2 output_options=foldmultilines=true
62838:
Output options for existing outputs are ignored.
```

(I can see `reconfigured` label in JDK 16 even if any output options are not accepted from jcmd - it may be bug)

File: src/hotspot/share/logging/logOutput.cpp (hunk 1 above)
Comment (YaSuenag) [reply src/hotspot/share/logging/logOutput.cpp]: Good catch! I fixed it in new commit.

**Review by dholmes-ora — APPROVED** [review]
Hi Yasumasa,

Update looks good.

So my comment about foldMultiLines could only be true for stdout/err and only in the future when it is enabled for those.

Thanks,
David

**Comment by YaSuenag** [comment]
Thanks David for your review!

> So my comment about foldMultiLines could only be true for stdout/err and only in the future when it is enabled for those.

I agree with you that the problem is not practical concern because the log which happens before argument processing is determinative, so I think we can excuse they are not folded. I'll check what logs can happen before argument processing.

**Pull request closed without merging**